Implement force push down for nested group by query

samarthjain opened the pull request:
@gianm @jihoonson - creating a PR as we discussed over the email. The pull request isn't complete by any means. I just wanted to get your feedback on the approach I have taken. For now, I have implemented the force push down feature. Once we have finalized on the over all approach, I will provide a patch for automatic push down in a separate patch. Thanks!

jihoonson commented:
Great! I'll take a look soon.

jihoonson commented:
Hi @samarthjain, I'll do my review today or tomorrow. Before that, I'd like to one question. Looks that a query context value decides pushing nested query execution or not. Does this mean all nested subqueries are pushed to historicals if the query is deeply nested?

samarthjain commented:
@jihoonson - Currently, when the user sets the push nested query execution flag on the outermost query, then the broker executes the entire nested query on the historical nodes. The broker then merges the results returned by the historical nodes as usual except that in this case, the rows returned by historical nodes have the structure of the outermost part of the query.

jihoonson commented:
Thanks. It sounds good for the first step. Will leave some comments soon!

jihoonson commented on processing/src/main/java/io/druid/query/groupby/GroupByQueryQueryToolChest.java:
```diff
@@ -215,13 +214,49 @@ public GroupByQueryQueryToolChest(
 
+  public static boolean shouldPushDownQuery(GroupByQuery q)
+  {
+    return QueryContexts.parseBoolean(q, GroupByQueryConfig.CTX_KEY_FORCE_PUSH_DOWN_NESTED_QUERY, false);
```
Can be simplified to `return q.getContextBoolean(GroupByQueryConfig.CTX_KEY_FORCE_PUSH_DOWN_NESTED_QUERY, false);` and then https://github.com/druid-io/druid/pull/5471/files#diff-76d2b703319f06a8103ebce8a58ef7fcR221 is not necessary.

jihoonson commented on processing/src/main/java/io/druid/query/groupby/GroupByQueryQueryToolChest.java:
```diff
@@ -162,47 +163,45 @@ public GroupByQueryQueryToolChest(
+          // Unset the push down nested query flag so that the historical doesn't erroneously end up pushing down the query itself
+          pushDownQueryBuilder.overrideContext(ImmutableMap.of(GroupByQueryConfig.CTX_KEY_FORCE_PUSH_DOWN_NESTED_QUERY, false));
+          GroupByQuery queryToPushDown = pushDownQueryBuilder.build();
+          innerQueryBuilder.setQueryToPushDown(queryToPushDown);
```
Would you elaborate more on this? `setQueryToPushDown()` means setting the outer query for the inner query?

samarthjain replied:
This was done to make sure that we can pass on the entire nested query over to the historical nodes. The role of the inner most query is only limited to figuring out what segments to query. But what gets executed on the historical nodes is this pushed down query which is nothing but the complete query. I think there is some improvement possible here. Right now, I am intercepting on DirectDruidClient to see whether the pushedDownQuery is present. If so, the DirectDruidClient gets hold of it and sends it over to the historical node. 

Relevant snippet in DirectDruidClient: https://github.com/druid-io/druid/pull/5471/files#diff-604c8c11338380d32a71a7621f031d2aR172

jihoonson commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByRowProcessor.java:
```diff
@@ -143,7 +156,7 @@ public boolean apply(Row input)
               Pair<Grouper<RowBasedKey>, Accumulator<AggregateResult, Row>> pair = RowBasedGrouperHelper.createGrouperAccumulatorPair(
                   query,
-                  true,
+                  true, //todo: samarth think about this attribute. I think this might be false in our case?
```
Not sure, but this comment might help.

```
If isInputRaw is true, transformations such as timestamp truncation and extraction functions have not
been applied to the input rows yet, for example, in a nested query, if an extraction function is being
applied in the outer query to a field of the inner query. This method must apply those transformations.
```

jihoonson commented on processing/src/main/java/io/druid/query/groupby/GroupByQuery.java:
```diff
@@ -102,6 +102,7 @@ public static Builder builder()
 
   private final boolean applyLimitPushDown;
   private final Function<Sequence<Row>, Sequence<Row>> postProcessingFn;
+  private final GroupByQuery pushedDownQuery;
```
Not sure this should be kept. When is this variable used?

jon-wei commented:
@samarthjain I think you can check out `GroupByLimitPushDownMultiNodeMergeTest` for an example of testing that kind of merging

jihoonson commented:
@jon-wei thanks. It looks a good example to test merging.

drcrallen commented:
Can this PR include more description in the master comment about what it is accomplishing and how it is accomplishing it?

samarthjain commented:
@jihoonson - I have updated pull request with review comments, did some clean up and added tests.

samarthjain commented:
The way nested query execution is implemented today, it executes the inner most query on the historical nodes with the outer queries being executed on the broker node. This can get problematic when the inner query groups using a high cardinality dimension, returning too many records for broker node to handle. One of the options that we have been internally testing and exploring is the capability to push down the complete nested query to the historical nodes. Each historical node then will execute the nested query primarily dealing with the segments it owns for that query. Because the number of records returned by each historical node would potentially be much smaller in this case, it would be less intensive for the broker to perform the final merge and aggregation. The broker though won't need to perform any more dimension or segment level filtering since it will be taken care of at the historical nodes itself.  Note that this way of distributing the aggregation to the historical nodes doesn't always return the same results as the final aggregation getting done on the broker node. However, there is a good set of cases (for ex - aggregating on dimensions that are used for hashing during ingestion) where this kind of push down logic will return the right results. I can get into this into more detail but the general idea was to leave the onus on the user to figure out if their data layout allows for this kind of push down.

This implementation provides user a way of forcing nested query execution through a query context variable. The next cut will focus on doing this automatically under the following conditions (credit to @gianm ) for clearly articulating the following:

1) The groupBy query granularity is equal to, or finer than, segment granularity;

and either:

2a) A time chunk uses HashBasedNumberedShardSpec, partitionDimensions is nonempty, the grouping dimension set contains all of the shard partitionDimensions, and there are no "extension" partitions (partitions with partitionNum >= partitions, which are created by ingest tasks that append data)

or:

2b) A time chunk uses SingleDimensionShardSpec and the grouping dimension set contains the shard dimension.

If Druid detects this it should push down the query automatically for that time chunk. There will be situations where the query can be pushed down for some time chunks but not others (for example: consider a data pipeline that loads data unpartitioned in realtime, and later has a reindexing job to partition historical data by some useful dimension). In this case, ideally the broker should be capable of pushing down the query for the time chunks where it can be correctly pushed down, and not pushing it down for others.

jihoonson commented:
@samarthjain thanks. I'll take another look.

jihoonson commented:
@samarthjain we are suffering from pretty frequent CI failures. If the CI for your PR fails, you first need to make sure the failure relates to any changes in your PR. This can be checked by running unit tests locally.

Usually the memory configuration for CI shouldn't be a problem. If you found something wrong in the CI configuration, please make another PR to fix it.

samarthjain commented:
@jihoonson - sorry for the late reply. I have cleaned up code, removed pom changes and added an integration test. I wanted to check with you how is the integration test triggered? How is the data needed for the test loaded? I looked at existing integration tests and couldn't figure out where that magic happens.

samarthjain commented:
{code}
[ERROR] before(io.druid.tests.query.ITNestedQueryPushDownTest)  Time elapsed: 403.943 s  <<< FAILURE!
io.druid.java.util.common.ISE: Max number of retries[10] exceeded for Task[wikipedia segment load]. Failing.
	at io.druid.tests.query.ITNestedQueryPushDownTest.before(ITNestedQueryPushDownTest.java:47)
{code}

Looks like the framework is not able to locate wikipedia data source which the new test needs. @jihoonson, the test passes for me locally after loading up the wikipedia datasource as mentioned on http://druid.io/docs/latest/tutorials/quickstart.html.  Would be great if you can let me know how to get load new data for the IT tests.

samarthjain commented:
@jihoonson - any insights on how to add and execute new integration tests for CI?

jihoonson commented:
@samarthjain I'll take a look today. Sorry for late review.

jihoonson commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByRowProcessor.java:
```diff
@@ -66,18 +69,28 @@
     for (int i = 0; i < query.getAggregatorSpecs().size(); i++) {
-      aggregatorFactories[i] = query.getAggregatorSpecs().get(i);
+      AggregatorFactory af = query.getAggregatorSpecs().get(i);
+      aggregatorFactories[i] = wasQueryPushedDown ? af.getCombiningFactory() : af;
```
Why is the combining factory used here?

samarthjain replied:
In case of nested query push down, the historicals are going to return aggregated results and the broker's job is to combine these aggregated results - hence the use of combining aggregator factory.

jihoonson replied:
Thanks.

Can we this kind of query rewriting for dimensions and aggregatorFactories in `GroupByQueryQueryToolChect.getResultsOnPushDown()`? It would be better because we can put necessary stuffs for nestedQueryPushDown together when `isNestedQueryPushDown()` is checked.

samarthjain commented:
Thanks for the detailed feedback, @jihoonson . I will review it and reply by tonight.

samarthjain commented:
> GroupByRowProcessor.process() is supposed to be called only when processing sub queries, but now it looks to be called when merging sub queries as well. I think it was for avoiding code duplication, but it can be improved by splitting them and extracting common code paths into utility methods.
The process method isn't doing any additional merges. The only change is to make it cognizant of the fact that the sub-query sequence it is processing in case of forcePushDown is now a result of running the pushed down nested query instead. 

>A groupBy v2 query requires to get a resource before query execution. We currently check there are enough number of merge buffers to execute a given query. Since historicals have processed only the inner-most query so far, they have always required a single merge buffer. However, they can process nested queries after this patch, resource preservation should be changed as well. Please see here for how the number of merge buffers is calculated for brokers.

Good point. I will make sure that historicals have same number of merge buffers as broker. Although from the first look at the code, it seems like it it should, but I will make sure and hopefully add a test for the same too.

 >If I understand correctly, once forceNestedQueryPushDown is set to true, the inner most query is passed from GroupByQueryQueryToolChest to DirectDruidClient to find which historicals should receive the query, and then that the pushed down nested query is sent from DirectDruidClient. I think this kind of value hijacking can cause many potential bugs. Instead, I suggest to add an interface to Query to find the dataSource of the inner-most query.

Thanks. Let me see how can I make this cleaner. I am not the biggest fan of the approach I took, either.

samarthjain commented on processing/src/main/java/io/druid/query/groupby/GroupByQueryQueryToolChest.java:
```diff
@@ -163,64 +165,154 @@ public GroupByQueryQueryToolChest(
+    ).withOverriddenContext(
+        ImmutableMap.<String, Object>of(
+            // the having spec shouldn't be passed down, so we need to convey the existing limit push down status
+            GroupByQueryConfig.CTX_KEY_APPLY_LIMIT_PUSH_DOWN, query.isApplyLimitPushDown(),
```
I actually copied this code/comment from https://github.com/druid-io/druid/blob/master/processing/src/main/java/io/druid/query/groupby/strategy/GroupByStrategyV2.java#L251.

samarthjain commented on processing/src/main/java/io/druid/query/dimension/PushDownQueryDimensionSpec.java:
```diff
@@ -0,0 +1,84 @@
+import io.druid.segment.DimensionSelector;
+import io.druid.segment.column.ValueType;
+
+public class PushDownQueryDimensionSpec implements DimensionSpec
```
The dimension name for this dimension spec is the the output name of the delegate. This dimension spec is only used in case of nested query push down.
`
 @Override
 public String getDimension()
{    // the dimension name is same as the output name.
    return delegate.getOutputName();
}
`

samarthjain commented:
>The integration test fails in my laptop as well. Also, when I tested the below query with my local cluster, I met this error.

Did you update the jar on the historicals too?

jihoonson commented:
@samarthjain yes, I tested in my local laptop and used the jar of the same version for all modules.

samarthjain commented:
Addressed the review comments you left, @jihoonson . I was able to get rid of the hacky approach I took for getting hold of the complete nested query in DirectDruidClient by using it directly higher in the stack in GroupByQueryToolChest and GroupByStrategyV2. Also  fixed the failure you ran into when using a multi-level nested query. To run the integration test on my laptop, I had to add a config file and supply it's path in the ConfigFileConfigProvider. The contents of the config file looked like this:
``
{   
   "broker_host": "127.0.0.1",
   "broker_port": "8082",
   "coordinator_host": "127.0.0.1",
   "coordinator_port": "8081",
   "middlemanager_host": "127.0.0.1",
   "middlemanager_port":"8090",
   "zookeeper_hosts": "127.0.0.1:2181"
}
``

I also had to change IntegrationTestingConfigProvider to use the ConfigFileConfigProvider and not DockerConfigProvider. 

@nishantmonu51 - would you happen to know how can I get my newly added integration test to run automatically in the pre-checkin builds?

jihoonson commented:
@samarthjain thanks for the update! I'll take another look.

>To run the integration test on my laptop, I had to add a config file and supply it's path in the ConfigFileConfigProvider. The contents of the config file looked like this:
{ "broker_host": "127.0.0.1", "broker_port": "8082", "coordinator_host": "127.0.0.1", "coordinator_port": "8081", "middlemanager_host": "127.0.0.1", "middlemanager_port":"8090", "zookeeper_hosts": "127.0.0.1:2181" }

Would you elaborate on why this configuration is needed? It's supposed to get addresses automatically in the integration tests.

> @nishantmonu51 - would you happen to know how can I get my newly added integration test to run automatically in the pre-checkin builds?

What do you mean by this? If you add an integration test, it should be run automatically on the travis. You can check it by checking the log on travis or running the integration test on your own.

jihoonson commented:
Also, please check the CI failure.

```
Results :
Failed tests: 
  GroupByLimitPushDownInsufficientBufferTest.testPartialLimitPushDownMergeForceAggs:605 expected:<MapBasedRow{timestamp=1970-01-01T00:00:00.000Z, event={dimA=mango, metA=190}}> but was:<MapBasedRow{timestamp=1970-01-01T00:00:00.000Z, event={dimA=zebra, metA=180}}>
Tests in error: 
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:6177 » ClassCast java.lang.Lo...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:6177 » ClassCast java.lang.Lo...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:6177 » ClassCast java.lang.Lo...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:6177 » ClassCast java.lang.Lo...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:6177 » ClassCast java.lang.Lo...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:6177 » ClassCast java.lang.Lo...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:6177 » ClassCast java.lang.Lo...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:6177 » ClassCast java.lang.Lo...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:6177 » ClassCast java.lang.Lo...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:6177 » ClassCast java.lang.Lo...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:6177 » ClassCast java.lang.Lo...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:6177 » ClassCast java.lang.Lo...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:6177 » ClassCast java.lang.Lo...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:6177 » ClassCast java.lang.Lo...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:6177 » ClassCast java.lang.Lo...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:6177 » ClassCast java.lang.Lo...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:6177 » ClassCast java.lang.Lo...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:6177 » ClassCast java.lang.Lo...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:6177 » ClassCast java.lang.Lo...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:6177 » ClassCast java.lang.Lo...
Tests run: 83353, Failures: 1, Errors: 20, Skipped: 62
```

samarthjain commented:
> What do you mean by this? If you add an integration test, it should be run automatically on the travis. You can check it by checking the log on travis or running the integration test on your own.

The core of the problem is that I used the wikiticker data set for my new integration test. I ended up assuming that the data set provided in http://druid.io/docs/latest/tutorials/quickstart.html will be available to run in the CI build environment too. But turns out, that isn't the case. At this point, it seems like I would have to emulate what other IT tests like ITIndexerTest are doing and load the data up in the setup phase of this new IT test. 

Looking at other IT tests like ITTwitterQueryTest and ITWikipediaQueryTest, I don't see where the data is first being loaded. My best guess is that the data is already available in some s3 bucket which then the docker cluster is downloading. For ex- I see lines like below in the sample-data.sql file - https://github.com/druid-io/druid/blob/master/integration-tests/docker/sample-data.sql

`INSERT INTO druid_segments (id,dataSource,created_date,start,end,partitioned,version,used,payload) VALUES ('twitterstream_2013-01-01T00:00:00.000Z_2013-01-02T00:00:00.000Z_2013-01-02T04:13:41.980Z_v9','twitterstream','2013-05-13T01:08:18.192Z','2013-01-01T00:00:00.000Z','2013-01-02T00:00:00.000Z',0,'2013-01-02T04:13:41.980Z_v9',1,'{\"dataSource\":\"twitterstream\",\"interval\":\"2013-01-.....
`

It would be nice if there was a way to access the twitter stream data used by these IT tests. That way I could have written tests for this feature with queries and expected results. Without access to the data, that isn't possible.

jihoonson commented:
@samarthjain ah, you're correct. Unfortunately, I also have no idea where the data is. @nishantmonu51 would you explain these things like where the data is and how to see/modify the input data?

jihoonson commented:
@samarthjain I found the location of test data is in the sql file. You can find it by searching bucket and key. Since the bucket is public, you should be able to download them. Once you download them, you can check data by using [DumpSegment tool](http://druid.io/docs/latest/operations/dump-segment.html).

jihoonson commented on processing/src/main/java/io/druid/query/groupby/GroupByQueryQueryToolChest.java:
```diff
@@ -203,24 +205,75 @@ public GroupByQueryQueryToolChest(
+  }
 
-      return groupByStrategy.processSubqueryResult(subquery, query, resource, finalizingResults);
+  public static GroupByQuery getNestedQueryToPushDown(GroupByQuery query)
```
Can be package-private. Also please annotate with `@VisibleForTesting`.

jihoonson replied:
This method is similar to [this code](https://github.com/druid-io/druid/blob/master/processing/src/main/java/io/druid/query/groupby/strategy/GroupByStrategyV2.java#L232-L254) in `GroupByStrategyV2.mergeResults()`, and the returned query of this method is immediately passed to that method. Is this needed?

jihoonson replied:
Or is this method supposed to return a subquery?

jihoonson commented on processing/src/main/java/io/druid/query/groupby/GroupByQueryQueryToolChest.java:
```diff
@@ -203,24 +205,75 @@ public GroupByQueryQueryToolChest(
+    ).withOverriddenContext(
+        ImmutableMap.of(
+            // the having spec shouldn't be passed down, so we need to convey the existing limit push down status
+            GroupByQueryConfig.CTX_KEY_APPLY_LIMIT_PUSH_DOWN, query.isApplyLimitPushDown(),
```
Should be 

```
        ImmutableMap.of(
            // the having spec shouldn't be passed down, so we need to convey the existing limit push down status
            GroupByQueryConfig.CTX_KEY_APPLY_LIMIT_PUSH_DOWN,
            query.isApplyLimitPushDown(),
            "finalize",
            false
        )
```

jihoonson commented on processing/src/main/java/io/druid/query/groupby/GroupByQueryQueryToolChest.java:
```diff
@@ -203,24 +205,75 @@ public GroupByQueryQueryToolChest(
-      return groupByStrategy.processSubqueryResult(subquery, query, resource, finalizingResults);
+  public static GroupByQuery getNestedQueryToPushDown(GroupByQuery query)
+  {
+    GroupByQuery queryToPushDown = new GroupByQuery(
```
`queryToPushDown` is redundant.

jihoonson commented on processing/src/main/java/io/druid/query/groupby/strategy/GroupByStrategyV2.java:
```diff
@@ -226,6 +226,7 @@ public boolean doMergeResults(final GroupByQuery query)
       }
     };
 
+    // Fudge timestamp, maybe.
```
Duplicated.

samarthjain commented:
@jihoonson - sorry for the delay in updating this pull request. I have been out on parental leave. I got the integration test working in my latest commit. Also addressed the code review comments.

samarthjain commented:
The processing module test passed fine for me locally.

jihoonson commented:
@samarthjain congratulations! I'll take another look!

jihoonson commented:
@samarthjain not really.. Sorry for the delayed review. I've been quite busy for our next minor release (0.12.2). I'll take a look in a few days.

jihoonson commented:
Hi @samarthjain, sorry for the delay. I'm looking at this PR again, and the size of this PR looks huge. Would you check that this PR contains only your changes?

samarthjain commented:
@jihoonson - the PR contains my changes only.

jihoonson commented:
@samarthjain thanks. It looks the huge patch size is because of test data. I think it's fine.

jihoonson commented on processing/src/main/java/io/druid/query/groupby/GroupByQueryConfig.java:
```diff
@@ -29,6 +29,8 @@
   public static final String CTX_KEY_FORCE_LIMIT_PUSH_DOWN = "forceLimitPushDown";
   public static final String CTX_KEY_APPLY_LIMIT_PUSH_DOWN = "applyLimitPushDown";
+  public static final String CTX_KEY_FORCE_PUSH_DOWN_NESTED_QUERY = "forcePushDownNestedQuery";
+  public static final String CTX_KEY_EXECUTING_NESTED_QUERY = "executinNestedQuery";
```
typo: `executinNestedQuery` -> `executingNestedQuery`

jihoonson commented on processing/src/main/java/io/druid/query/groupby/GroupByQueryQueryToolChest.java:
```diff
@@ -162,10 +162,11 @@ public GroupByQueryQueryToolChest(
       Map<String, Object> context
   )
   {
+    if (isNestedQueryPushDown(query, groupByStrategy)) {
```
I believe that once `isNestedQueryPushDown(query, groupByStrategy)` returns false, it never changes no matter how many times we call it in this method. Since `mergeGroupByResults()` can call internally itself again if `dataSource` is a `queryDataSource` (https://github.com/apache/incubator-druid/pull/5471/files#diff-631de2c7bba9484c3ffd4a58b4cf691eR193), this might be checked again and again redundantly. I would suggest to add new methods which is called in `mergeGroupByResults()` like below:

```
mergeGroupByResults() -> mergeGroupByResultsWithoutPushDown() -> mergeGroupByResultsWithoutPushDown() if needed ...
                     |-> mergeGroupByResultsWithPushDown()
```

jihoonson commented on processing/src/main/java/io/druid/query/groupby/GroupByQueryQueryToolChest.java:
```diff
@@ -516,4 +540,5 @@ public boolean apply(DimensionSpec input)
         }
     );
   }
+
```
Please remove unnecessary change.

jihoonson commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByRowProcessor.java:
```diff
@@ -66,18 +70,28 @@
-    final GroupByQuery query = (GroupByQuery) queryParam;
+    GroupByQuery queryWithRenamedDimensions = null;
+    if (wasQueryPushedDown) {
+      List<DimensionSpec> newDims = new ArrayList<>(((GroupByQuery) queryParam).getDimensions().size());
```
nit: can be simpler.

```java
      final List<DimensionSpec> newDims = ((GroupByQuery) queryParam).getDimensions()
                                                                     .stream()
                                                                     .map(RenamedDimensionSpec::new)
                                                                     .collect(Collectors.toList());
```

jihoonson commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByRowProcessor.java:
```diff
@@ -98,7 +112,18 @@
+          public boolean apply(Row input)
+          {
+            rowSupplier.set(input);
+            return true; // nothing to filter since it has already been done on the push down
```
If this is the case, I believe it's much faster to call `final AggregateResult retVal = rows.accumulate(AggregateResult.ok(), accumulator);` at https://github.com/apache/incubator-druid/pull/5471/files#diff-dfac707f8edd4ee002b9f3372cfc2699R192.

jihoonson commented on server/src/main/java/io/druid/client/CachingClusteredClient.java:
```diff
@@ -235,6 +237,9 @@ public CachingClusteredClient(
       this.downstreamQuery = query.withOverriddenContext(makeDownstreamQueryContext());
+      this.intervals = query instanceof BaseQuery
+                                        ? BaseQuery.getIntervals((BaseQuery) query)
+                                        : query.getIntervals();
```
formatting:

```java
      this.intervals = query instanceof BaseQuery ?
                       BaseQuery.getIntervals((BaseQuery) query) :
                       query.getIntervals();
```

jihoonson commented on server/src/main/java/io/druid/server/QueryLifecycle.java:
```diff
@@ -249,8 +249,7 @@ public QueryResponse execute()
 
-    final Sequence res = QueryPlus.wrap(baseQuery)
-                                  .withIdentity(authenticationResult.getIdentity())
+    final Sequence res = QueryPlus.wrap(baseQuery).withIdentity(authenticationResult.getIdentity())
```
It looks to be the right formatting before being changed.

jihoonson commented on processing/src/main/java/io/druid/query/BaseQuery.java:
```diff
@@ -257,4 +257,17 @@ public int hashCode()
+   *
+   * @return intervals to query. For nested queries, return the intervals for the inner most query.
+   */
+  public static List<Interval> getIntervals(BaseQuery query)
```
Since every query implementation extends `BaseQuery`, I think it's better to add this method to `Query` which is the higher interface than `BaseQuery`. What do you think about adding a method to `Query` like below:

```java
default List<Interval> getIntervalsOfInnerMostQuery()
  {
    if (getDataSource() instanceof QueryDataSource) {
      //noinspection unchecked
      return ((QueryDataSource) getDataSource()).getQuery().getIntervalsOfInnerMostQuery();
    } else {
      return getIntervals();
    }
  }
```

jihoonson commented on server/src/main/java/io/druid/server/coordination/ServerManager.java:
```diff
@@ -109,6 +110,14 @@ public ServerManager(
     this.serverConfig = serverConfig;
   }
 
+  private DataSource getTableDataSource(DataSource dataSource)
```
nit: `getInnerMostDataSource()` sounds more intuitive.

samarthjain commented:
@jihoonson - I finally got around to addressing your review comments in this latest push. Something got messed up after the recent migration to incubator-apache and it messed up my local git history. So unfortunately I had to do a hard reset of my previous commits and a force push for this pull request. I didn't want to create a new request since it would have taken away all the valuable conversation we had here. Hope it's ok.

samarthjain commented:
Not sure why the travis build is complaining. Seems to be building fine locally. @jihoonson - is there a way for you to retry the build?

samarthjain commented:
Ok, looks like there was a rebase issue with my branch. @jihoonson - would be awesome if you can take a look when you get a chance.

samarthjain commented:
@jihoonson - I see that a 0.13.0 release is being planned. It would be great if this feature can be added to the 0.13.0 release.

jihoonson commented:
Hi @samarthjain! Thank you for updating this PR. I'll take a look soon, probably at least in next week. Unfortunately, the feature for 0.13.0 is frozen, but I'm scheduling this to 0.13.1. Thanks!

jihoonson commented on processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/GroupByRowProcessor.java:
```diff
@@ -152,7 +152,9 @@ public ByteBuffer get()
-    final AggregateResult retVal = filteredSequence.accumulate(AggregateResult.ok(), accumulator);
+    final AggregateResult retVal = wasQueryPushedDown
+                                   ? rows.accumulate(AggregateResult.ok(), accumulator)
+                                   : filteredSequence.accumulate(AggregateResult.ok(), accumulator);
```
Is this because the filters are already processed when `wasQueryPushedDown` is true? Then, please add a comment about it here.

jihoonson replied:
Also, I think we can clarify this part a bit more. For example, all filter-related variables like `filteredSequence`, `filterMatcher`, `filter`, etc don't have to be initialized. Would you please export this part as a separate method and skip calling it if `wasQueryPushedDown` is true?

jihoonson commented on processing/src/test/java/org/apache/druid/query/groupby/NestedQueryPushDownTest.java:
```diff
@@ -0,0 +1,1059 @@
+    );
+    GroupByQuery rewrittenQuery = ((GroupByQueryQueryToolChest) toolChest).rewriteNestedQueryForPushDown(nestedQuery);
+    // Broker executes this code and hence has
+    return strategy.applyPostProcessing(strategy.processSubqueryResult(
```
Can we call `GroupByQueryQueryToolChest.mergeResults()` instead of calling this directly? I think it would be better because the unit test would execute the same logic with the production code.

samarthjain commented:
Thanks for your patient reviews, @jihoonson. I have updated the PR with the requested changes. Travis build seems to be timing out.

samarthjain commented:
@jihoonson - does this look good to merge now?

samarthjain commented:
CI looks good. Thanks for restarting it and all your reviews, @jihoonson. Let me know if you need to me to squash the commits.

jihoonson commented:
@samarthjain nice! You don't have to squash commits in most cases. I'll merge this shortly.

leventov commented on processing/src/test/java/org/apache/druid/query/groupby/NestedQueryPushDownTest.java:
```diff
@@ -0,0 +1,1038 @@
+    Assert.assertEquals(outputNameAgg, rewrittenQuery.getAggregatorSpecs().get(0).getName());
+  }
+
+  private List<QueryRunner<Row>> getRunner1()
```
Unused method

leventov commented on processing/src/test/java/org/apache/druid/query/groupby/NestedQueryPushDownTest.java:
```diff
@@ -0,0 +1,1038 @@
+    return runners;
+  }
+
+  private List<QueryRunner<Row>> getRunner2()
```
Unused method
